### PR TITLE
Enable preloading of a model

### PR DESCRIPTION
**Pull request by hesther**
As discussed, here is an (initial, unfinished) PR of enabling model preloading. To test, run a script like:

```python
import chemprop
import time

smiles = [
    ['CC(C)C(C)(C)C'],
    ['Clc1ccc(c(Cl)c1Cl)c2c(Cl)cc(Cl)c(Cl)c2Cl'],
]

# Build arguments
arguments = [
    '--test_path', '/dev/null',
    '--preds_path', '/dev/null',
    '--checkpoint_paths', 'model.pt', #change to your model file or files, or use checkpoint_dir if needed
]

# Parse arguments
args = chemprop.args.PredictArgs().parse_args(arguments)
print(args)

# Run predictions individually
time1=time.time()
for i in range(100):
    preds = chemprop.train.make_predictions(args=args, smiles=smiles)
time2=time.time()
print(time2-time1)
print(preds)

# Run predictions together
time1=time.time()
model_args = chemprop.train.load_model(args=args)
for i in range(100):
    preds = chemprop.train.make_predictions(args=args, smiles=smiles, model_args=model_args)
time2=time.time()
print(time2-time1)
print(preds)

```

I restructured `make_predictions()`, splitting it up to smaller functions, which are also more easy to work with. I exposed `load_model()` in `__init__.py`, but not the other functions (which we might want to do, to let users e.g. also preload data!)

What do you think? This version is still missing docstrings and variable types, I just wanted to hear your thoughts before I go any further.

One downside of the current implementation is that we hold all models in an ensemble in memory, even if we don't preload. I can change that, but it will make the code slightly less pretty.

### CODE REVIEW

File: chemprop/train/make_predictions.py [hunk 1]
```diff
@@ -31,16 +30,25 @@ def make_predictions(args: PredictArgs, smiles: List[List[str]] = None) -> List[
...
-        set_extra_atom_fdim(train_args.atom_features_size)
-
-    if args.bond_features_path is not None:
-        set_extra_bond_fdim(train_args.bond_features_size)
+    # Load model and scalers
+    models=(load_checkpoint(checkpoint_path, device=args.device) for checkpoint_path in args.checkpoint_paths)
+    scalers=(load_scalers(checkpoint_path) for checkpoint_path in args.checkpoint_paths)
+    if not generator:
+        models=list(models)
+        scalers=list(scalers)
```
Comment (mliu49) [inline chemprop/train/make_predictions.py:38]: Could you add spaces around the assignment `=` signs? (No spaces around keyword argument `=`)

File: chemprop/train/make_predictions.py [hunk 2]
```diff
@@ -1,28 +1,27 @@
...
 from chemprop.args import PredictArgs, TrainArgs
-from chemprop.data import get_data, get_data_from_smiles, MoleculeDataLoader, MoleculeDataset
+from chemprop.data import get_data, get_data_from_smiles, MoleculeDataLoader, MoleculeDataset, StandardScaler
 from chemprop.utils import load_args, load_checkpoint, load_scalers, makedirs, timeit, update_prediction_args
 from chemprop.features import set_extra_atom_fdim, set_extra_bond_fdim, set_reaction, set_explicit_h
+from chemprop.models import MoleculeModel
 
-@timeit()
-def make_predictions(args: PredictArgs, smiles: List[List[str]] = None) -> List[List[Optional[float]]]:
+def load_model(args: PredictArgs, generator: bool =False):
```
Comment (mliu49) [inline chemprop/train/make_predictions.py:15]: Could you add a space after the `=`?

File: chemprop/train/make_predictions.py [hunk 3]
```diff
@@ -179,6 +222,40 @@ def make_predictions(args: PredictArgs, smiles: List[List[str]] = None) -> List[
...
+        
+    set_features(args, train_args)
+    
+    full_data, test_data, test_data_loader, full_to_valid_indices = load_data(args, smiles)
+    
+    # Edge case if empty list of smiles is provided
+    if len(test_data) == 0:
+        return [None] * len(full_data)
+    
+    avg_preds=predict_and_save(args, train_args, test_data, task_names, num_tasks, test_data_loader, full_data, full_to_valid_indices, models, scalers)
```
Comment (mliu49) [inline chemprop/train/make_predictions.py:257]: Spaces around `=`

**Review by mliu49 — COMMENTED** [review]
I added some minor code style comments. The changes look good otherwise, although I have note tested it.

File: chemprop/train/make_predictions.py (hunk 1 above)
Comment (hesther) [reply chemprop/train/make_predictions.py:38]: Done!

File: chemprop/train/make_predictions.py (hunk 2 above)
Comment (hesther) [reply chemprop/train/make_predictions.py:15]: Done!

File: chemprop/train/make_predictions.py (hunk 3 above)
Comment (hesther) [reply chemprop/train/make_predictions.py:257]: Done!

**Comment by hesther** [comment]
@mliu49 Is there anything else we would want to include, or is this ready to push (after receiving a positive review)?

**Comment by mliu49** [comment]
Sorry for the delay, the only thing left is whether I should also test the changes? If so, could you provide a dummy model or instructions for making one?

**Comment by hesther** [comment]
You could test with something like (which makes 100 predictions with and without preloading):

```python
import chemprop
import time

smiles = [
    ['CC(C)C(C)(C)C'],
    ['Clc1ccc(c(Cl)c1Cl)c2c(Cl)cc(Cl)c(Cl)c2Cl'],
]

# Build arguments
arguments = [
    '--test_path', '/dev/null',
    '--preds_path', '/dev/null',
    '--checkpoint_paths', 'model.pt', #change to your model file or files, or use checkpoint_dir if needed
]

# Parse arguments
args = chemprop.args.PredictArgs().parse_args(arguments)
print(args)

# Run predictions individually
time1=time.time()
for i in range(100):
    preds = chemprop.train.make_predictions(args=args, smiles=smiles)
time2=time.time()
print(time2-time1)
print(preds)

# Run predictions together
time1=time.time()
model_args = chemprop.train.load_model(args=args)
for i in range(100):
    preds = chemprop.train.make_predictions(args=args, smiles=smiles, model_args=model_args)
time2=time.time()
print(time2-time1)
print(preds)

```

**Comment by hesther** [comment]
I just edited the instructions in the comment above, since once of the variable names changes. Now it should work (sorry if you already ran it!)

This should work with any flavour of chemprop, e.g. ensembles, multiple molecules, reactions, etc

**Comment by mliu49** [comment]
Thanks! I built a model using `tests/data/regression.csv`, and only saw a ~3s speedup (out of 80s total) using the test code you provided. Do you think that's because the model was small/fast to load?

Other than that, everything worked, and I think this can be merged!

**Review by mliu49 — APPROVED** [review]

**Comment by hesther** [comment]
Mmh good question. I good speedups of factor 2-5 (e.g. for a single model with data/delaney, I went from 4.7s to 1.0s). I will try to reproduce your results.

**Comment by hesther** [comment]
@mliu49, I tried to reproduce your results and ran:
```python train.py --data_path tests/data/regression.csv --dataset_type regression --epochs 2 --save_dir test_model```
```python test.py```
(with test.py being the code posted above), and got 4.2 seconds (without preloading) and 0.9 second (with preloading) on my old laptop, so I wonder how you can get up to 80 seconds?

What system (CPU/GPU, OS, ...) did you test on?

**Comment by hesther** [comment]
If this was on a cluster, I wonder whether you had 76s overhead + 4s calculating (no preloading) and 76s overhead + 1 s calculating (preloading), which would be the same as I had ... e.g. supercloud takes many seconds to load packages. Could you increase e.g. to 100 iterations, so we can see how much overhead we have?

**Comment by mliu49** [comment]
I ran on a MacBook Pro (2.3 GHz 8-Core Intel Core i9), interactively in Jupyter, with PyTorch 1.7.1 cpu-only.

I was also surprised that it was taking so long for me.

**Comment by hesther** [comment]
And CPU, Jupyter:  4.9s vs 1.3s, so it isn't the notebook playing up.

**Comment by mliu49** [comment]
After more testing, I've confirmed that it's environment related, and is most likely due to the Python version.

My original environment was Python 3.9, and I also confirmed that Python 3.8 gave the same result. However, I was able to get times of 4s and 1s using Python 3.7. The environments also had different PyTorch versions, which is why I'm not 100% confident.

My hypothesis is that the cause is the change in multiprocessing on MacOS introduced in Python 3.8 (which has caused issues before). There were two other supporting observations:
- The logs showed descriptastorus warnings on every prediction in Python 3.8/3.9, suggesting that it was being reloaded every time, which may be the main contributor to runtime
- I was unable to run the test as a normal Python script in Python 3.8/3.9 due to multiprocessing errors, although it worked in Jupyter

In conclusion, this PR is fine, but perhaps we should consider limiting Python version support (at least for MacOS).

**Comment by hesther** [comment]
I very much support limiting Python versions for MacOS, because as I had said before (issue #115), everything above Python 3.7 causes an error with the Global variables

**Comment by hesther** [comment]
I will go ahead and merge this now
